### DISCUSSION TITLE
Allow build to export as an importable package

### DISCUSSION THREAD

**Discussion by primetoxinz** [livebud/bud#58] [Ideas]
Hey, team. Bud is looking awesome and very useful!
I am looking to use bud for generating internal administrative UIs for my companies microservice infrastructure - I want these UIs to be served by an existing router, rather than by its own binary.

So I'm wondering how difficult it would be for `bud build` to have an option to output the .app code to an actual package in a project that exposes the `http.Handler` for adding to an existing `http.Server` rather than using bud`s

**Comment by matthewmueller** [comment]
Hey @primetoxinz thanks for your interest in Bud! 

In your case, I'd try build a bud app as a separate module, then creating a `server.go` at the root level to expose the web server. 

```
budapp/
  bud/
    .app/
       web/
         web.go
  go.mod
  server.go
```

This `server.go` would reach into `bud/.app/web` and initialize the web server. 

Inside your `server.go`, you'd have something like this (untested):

```go
package budapp

import (
  "budapp/bud/.app/web"
)

func Load() (*web.Server, error) {
  // TODO: wire up the dependencies needed for New
  return web.New(...)
}
```

Once you have this setup, you can import `budapp` into your admin UI like you would any other go module.

This will some manual work, because you'll need to wire the dependencies yourself. Have a look at `bud/.app/program/program.go` to  see how the dependency injection framework did the wiring. Just a warning that this is a private API and bud is a fast moving project right now, so expect some breaking changes here. Whenever you encounter them, feel free to ping me for help.

In the future, I plan to expose the `*web.Server` as a fully loaded type, because in addition to your use case of bringing bud into existing apps, it will be useful for writing functional tests. If you'd like to contribute to this effort, please let me know!
